REGRESSION: Space bar for page down works intermittently

rr-codes opened the pull request:
#### f52fe75ad080437a7e5705f3ae3632797d537818
<pre>
REGRESSION: Space bar for page down works intermittently
<a href="https://bugs.webkit.org/show_bug.cgi?id=258434">https://bugs.webkit.org/show_bug.cgi?id=258434</a>
rdar://109638580

Reviewed by Simon Fraser.

Keyboard scrolls do not begin if a user scroll is &quot;in progress&quot;. This uncovered a bug with normal
scrolling, which is that after performing a trackpad scroll with momentum, then tapping with two
fingers, the `isUserScrollInProgress` bool was still `true`, even though at this point the scroll
has stopped and so it should be `false`.

When performing the scroll with momentum, the following sequence of wheel events phases happen:
1. Begin -&gt; `isUserScrollInProgress = true`
2. End -&gt; `isUserScrollInProgress = false`
3. Momentum Begin -&gt; `isUserScrollInProgress = true`
4. Momentum End -&gt; `isUserScrollInProgress = false`

Then, when tapping with two fingers, these wheel event phases happen:
5. May Begin -&gt; `isUserScrollInProgress = true`
6. Cancelled -&gt; No action.

The root problem is that when receiving a wheel event whose phase is &quot;Cancelled&quot;, `isUserScrollInProgress`
is not being set to `false`.

This is because in `RemoteScrollingCoordinator`, the `m_nodesWithActiveUserScrolls` set does not get
cleared when cancelling, which is what `isUserScrollInProgress` depends upon. This set comes from
the state in `RemoteScrollingUIState`.

For wheel events that have a phase that isn&apos;t &quot;Cancelled&quot;, `ScrollingTree::setUserScrollInProgressForNode`
gets called, and `m_treeState.nodesWithActiveUserScrolls` has nodes either added or removed. Then,
`RemoteScrollingCoordinatorProxyMac::scrollingTreeNode{WillStart,DidEnd}Scroll` gets called, which
updates the state of `RemoteScrollingUIState` and sends an update.

However, for &quot;Cancelled&quot; wheel events, `ScrollingTree::clearNodesWithUserScrollInProgress` gets called,
which clears `m_treeState.nodesWithActiveUserScrolls`, but does *not* update the corresponding state of
`RemoteScrollingUIState` or send an update, so `m_nodesWithActiveUserScrolls` still has the stale nodes.

To fix, just update the `RemoteScrollingUIState` state when clearing nodes like we do when adding
or removing them.

* LayoutTests/fast/scrolling/mac/keyboard-scroll-after-wheel-event-cancel-expected.txt: Added.
* LayoutTests/fast/scrolling/mac/keyboard-scroll-after-wheel-event-cancel.html: Added.
* Source/WebCore/page/scrolling/ScrollingTree.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteScrollingUIState.cpp:
(WebKit::RemoteScrollingUIState::clearNodesWithActiveUserScroll):
* Source/WebKit/Shared/RemoteLayerTree/RemoteScrollingUIState.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.h:
(WebKit::RemoteScrollingCoordinatorProxy::clearNodesWithUserScrollInProgress):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingTree.cpp:
(WebKit::RemoteScrollingTree::clearNodesWithUserScrollInProgress):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingTree.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingCoordinatorProxyMac.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingCoordinatorProxyMac.mm:
(WebKit::RemoteScrollingCoordinatorProxyMac::clearNodesWithUserScrollInProgress):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.mm:
(WebKit::RemoteScrollingTreeMac::clearNodesWithUserScrollInProgress):

Canonical link: <a href="https://commits.webkit.org/265602@main">https://commits.webkit.org/265602@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e680cc867830ecfc2942f396a8618a86062c85e2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/11093 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/11305 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/11628 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/12743 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/10586 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/11107 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/13682 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/11288 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/13493 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/11253 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/12142 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/9355 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/13147 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/9435 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/10030 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/17234 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/10506 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/10184 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/13405 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/10618 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/8705 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/9789 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/9921 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2726 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/14062 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/10472 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/page/scrolling/ScrollingTree.h:
```diff
@@ -141,6 +141,7 @@ class ScrollingTree : public ThreadSafeRefCounted<ScrollingTree> {
 #endif
     virtual void scrollingTreeNodeWillStartScroll(ScrollingNodeID) { }
     virtual void scrollingTreeNodeDidEndScroll(ScrollingNodeID) { }
+    virtual void scrollingTreeNodeWillCancelScroll() { }
```
This naming isn't right, since this action isn't associated with a specific `ScrollingNodeID`; it's not an analogue of `scrollingTreeNodeDidEndScroll()`.

Why not just make `clearNodesWithUserScrollInProgress()` virtual, and override in `RemoteScrollingTree`?

rr-codes replied:
Will do

aprotyas commented on Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.mm:
```diff
@@ -90,6 +90,13 @@
+void RemoteScrollingTreeMac::clearNodesWithUserScrollInProgress()
+{
+    RunLoop::main().dispatch([protectedThis = Ref { *this }, this] {
+        RemoteScrollingTree::clearNodesWithUserScrollInProgress();
```
hmm.. `RemoteScrollingTree::clearNodesWithUserScrollInProgress()` is not a static function, can we call it like this?

rr-codes replied:
yes; it's calling the super method

aprotyas replied:
oh! My bad

aprotyas commented on LayoutTests/fast/scrolling/mac/keyboard-scroll-after-wheel-event-cancel.html:
```diff
@@ -0,0 +1,78 @@
+<body onload="runTest()">
+    <div style="height: 5000px;"></div>
+</body>
+</html>
```
optional nit: trailing new line

smfr commented on Source/WebKit/Shared/RemoteLayerTree/RemoteScrollingUIState.cpp:
```diff
@@ -96,4 +96,13 @@ void RemoteScrollingUIState::removeNodeWithActiveUserScroll(WebCore::ScrollingNo
         m_changes.add(Changes::UserScrollNodes);
 }
 
+void RemoteScrollingUIState::clearNodesWithActiveUserScroll()
```
Don't you need to call the base class?

rr-codes replied:
The other `RemoteScrollingUIState` methods don't seem to call any base class?

smfr replied:
Sorry, I meant to comment on `RemoteScrollingTree::clearNodesWithUserScrollInProgress()`

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingTree.cpp:
```diff
@@ -119,6 +119,12 @@ void RemoteScrollingTree::scrollingTreeNodeDidEndScroll(ScrollingNodeID nodeID)
         m_scrollingCoordinatorProxy->scrollingTreeNodeDidEndScroll(nodeID);
 }
 
+void RemoteScrollingTree::clearNodesWithUserScrollInProgress()
```
Does this need to call the base class?

cdumez commented on LayoutTests/fast/scrolling/mac/keyboard-scroll-after-wheel-event-cancel.html:
```diff
@@ -0,0 +1,78 @@
+    <script src="../../../resources/ui-helper.js"></script>
+    <script src="../../../resources/js-test.js"></script>
+    <script>
+        jsTestIsAsync = true;
```
This test is missing a `description()` call to explain what the test does.

rr-codes committed:
REGRESSION: Space bar for page down works intermittently
https://bugs.webkit.org/show_bug.cgi?id=258434
rdar://109638580

Reviewed by Simon Fraser.

Keyboard scrolls do not begin if a user scroll is "in progress". This uncovered a bug with normal
scrolling, which is that after performing a trackpad scroll with momentum, then tapping with two
fingers, the `isUserScrollInProgress` bool was still `true`, even though at this point the scroll
has stopped and so it should be `false`.

When performing the scroll with momentum, the following sequence of wheel events phases happen:
1. Begin -> `isUserScrollInProgress = true`
2. End -> `isUserScrollInProgress = false`
3. Momentum Begin -> `isUserScrollInProgress = true`
4. Momentum End -> `isUserScrollInProgress = false`

Then, when tapping with two fingers, these wheel event phases happen:
5. May Begin -> `isUserScrollInProgress = true`
6. Cancelled -> No action.

The root problem is that when receiving a wheel event whose phase is "Cancelled", `isUserScrollInProgress`
is not being set to `false`.

This is because in `RemoteScrollingCoordinator`, the `m_nodesWithActiveUserScrolls` set does not get
cleared when cancelling, which is what `isUserScrollInProgress` depends upon. This set comes from
the state in `RemoteScrollingUIState`.

For wheel events that have a phase that isn't "Cancelled", `ScrollingTree::setUserScrollInProgressForNode`
gets called, and `m_treeState.nodesWithActiveUserScrolls` has nodes either added or removed. Then,
`RemoteScrollingCoordinatorProxyMac::scrollingTreeNode{WillStart,DidEnd}Scroll` gets called, which
updates the state of `RemoteScrollingUIState` and sends an update.

However, for "Cancelled" wheel events, `ScrollingTree::clearNodesWithUserScrollInProgress` gets called,
which clears `m_treeState.nodesWithActiveUserScrolls`, but does *not* update the corresponding state of
`RemoteScrollingUIState` or send an update, so `m_nodesWithActiveUserScrolls` still has the stale nodes.

To fix, just update the `RemoteScrollingUIState` state when clearing nodes like we do when adding
or removing them.

* LayoutTests/fast/scrolling/mac/keyboard-scroll-after-wheel-event-cancel-expected.txt: Added.
* LayoutTests/fast/scrolling/mac/keyboard-scroll-after-wheel-event-cancel.html: Added.
* Source/WebCore/page/scrolling/ScrollingTree.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteScrollingUIState.cpp:
(WebKit::RemoteScrollingUIState::clearNodesWithActiveUserScroll):
* Source/WebKit/Shared/RemoteLayerTree/RemoteScrollingUIState.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.h:
(WebKit::RemoteScrollingCoordinatorProxy::clearNodesWithUserScrollInProgress):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingTree.cpp:
(WebKit::RemoteScrollingTree::clearNodesWithUserScrollInProgress):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingTree.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingCoordinatorProxyMac.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingCoordinatorProxyMac.mm:
(WebKit::RemoteScrollingCoordinatorProxyMac::clearNodesWithUserScrollInProgress):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.mm:
(WebKit::RemoteScrollingTreeMac::clearNodesWithUserScrollInProgress):

Canonical link: https://commits.webkit.org/265602@main
